### PR TITLE
Migrate CI to use swiftlang / SwiftNIO common GitHub Actions.

### PR DESCRIPTION
**Pull request by rnro**
### Motivation:

* Reduce duplication
* Centralise boilerplate changes when new Swift versions are picked up.
* Benefit from centralised work to add new linting / test infrastructure.

### Modifications:

Changes of note:
* Use soundness checks from swiftlang/github-workflows.
* Define a gRPC-specific soundness check which retains bespoke license-checking code for .swift files as the gRPC header style is very different to most templates and checks that generated code is up-to-date.

### Result:

More test, linting, formatting coverage. More common CI with other Swift on Server projects.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -17,3 +17,7 @@ In order to protect both you and ourselves, you will need to sign the
 
 Please see the [main gRPC repository](https://github.com/grpc/grpc) for
 more information about gRPC.
+
+### Run CI checks locally
+
+You can run the Github Actions workflows locally using [act](https://github.com/nektos/act) or in some cases calling scripts directly. For detailed steps on how to do this please see [https://github.com/swiftlang/github-workflows?tab=readme-ov-file#running-workflows-locally](https://github.com/swiftlang/github-workflows?tab=readme-ov-file#running-workflows-locally).
```
Comment (glbrntt) [inline CONTRIBUTING.md:23]:
```suggestion
You can run the GitHub Actions workflows locally using [act](https://github.com/nektos/act) or in some cases calling scripts directly. For detailed steps on how to do this please see [https://github.com/swiftlang/github-workflows?tab=readme-ov-file#running-workflows-locally](https://github.com/swiftlang/github-workflows?tab=readme-ov-file#running-workflows-locally).
```

File: Examples/hello-world/Protos.proto [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2015 gRPC authors.
```
Comment (glbrntt) [inline Examples/hello-world/Protos.proto:1]: Not sure what's happened here, but this is meant to be a symlink from `Examples/hello-world/Protos/HelloWorld.proto` to `dev/protos/upstream/grpc/examples/helloworld.proto`. Could you fix it?

**Review by glbrntt — CHANGES_REQUESTED** [review]
Thanks @rnro -- couple of small things to fix beyond what we already discussed.

File: CONTRIBUTING.md (hunk 1 above)
Comment (rnro) [reply CONTRIBUTING.md]: 🙏 It's a good job I haven't already pasted this all over the place...

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      linux_nightly_main_arguments_override: "--explicit-target-dependency-import-check error -Xswiftc -require-explicit-sendable"
+
+  integration-tests:
+    name: Integration tests
+    uses: apple/swift-nio/.github/workflows/swift_matrix.yml@main
+    with:
+      name: "Integration Tests"
+      matrix_linux_5_9_enabled: false
+      matrix_linux_5_10_enabled: false
+      matrix_linux_command: "./dev/integration_tests.sh"
```
Comment (glbrntt) [inline .github/workflows/main.yml:27]: This job is failing because this script doesn't exist 🙃

**Review by glbrntt — CHANGES_REQUESTED** [review]
We're missing the integration tests script but this looks good otherwise

File: .github/workflows/main.yml (hunk 3 above)
Comment (rnro) [reply .github/workflows/main.yml]: That's true! As far as I can tell this repo doesn't have integration tests beyond benchmarks so I've deleted this block.

**Review by glbrntt — APPROVED** [review]
Fab, thanks @rnro!

**Pull request merged**